src/QuadForm/Quad/ZGenusRep.jl

```diff
@@ -0,0 +1,1030 @@
+end
+
+###############################################################################
+#
+#  User interface
+#
+###############################################################################
+
+@doc raw"""
+    is_maximal_integral_bilinear(G::ZZGenus, p::ZZRingElem) -> Bool
+
+Given a genus `G` of integral $\mathbb{Z}$-lattices and a prime number
+`p`, return whether the lattice $L_p$ is maximal integral in
+$\mathbb{Q}L_p$, for the bilinear form on $L_p$.
+
+Note that this does not depend on a choice of a lattice $L$ in $G$,
+and this property can be directly check from the genus symbol.
+"""
+function is_maximal_integral_bilinear(G::ZZGenus, p::ZZRingElem)
+  @req is_integral(G) "Lattices in the genus must be globally integral"
+  sym = symbol(local_symbol(G, p))
+
+  # If the local lattice is unimodular, then the bilinear form on the
+  # discriminant group `q_p` is trivial and it has no isotropic vector.
+  #
+  # If the p-adic level `l = p^k` is greater or equal than p^2, then the
+  # bilinear form on `p^(k-1) * q_p` is totally isotropic, and `q_p` has an
+  # isotropic vector.
+  if sym[end][1] == 0
+    return true
+  elseif sym[end][1] >= 2
+    return false
+  end
+
+  # If the Jordan constituent of scale p has rank 1 then there is
+  # no isotropic vector for the bilinear form on the discriminant group.
+  #
+  # If it has rank 3 then as a consequence of Chevalley's theorem
+  # the bilinear form on the discriminant group has an isotropic vector.
+  if sym[end][2] == 1
+    return true
+  elseif sym[end][2] >= 3
+    return false
+  end
+
+  # For the rank 2, there are only two possible normal forms for the
+  # bilinear form on the discriminant group.
+  # In the case p=2, they both admit an isotropic vector
+  p == 2 && return false
+
+  # For the odd case, one of them has an isotropic vector but
+  # not the other one
+  sym[end][3] == jacobi_symbol(-1, p) && return false
+  return true
+end
+
+@doc raw"""
+    _is_isotropic(G::ZZGenus, p::ZZRingElem) -> Bool
+
+Return whether a lattice in the genus $G$ is isotropic locally at $p$.
+"""
+function _is_isotropic(G::ZZGenus, p::ZZRingElem)
+  @req is_prime(p) "Number p must be prime"
+  d = det(G)
+  n = rank(G)
+  if d == 0
+    return true
+  elseif n <= 1
+    return false
+  elseif n == 2
+    return is_local_square(-d, p)
+  elseif n == 3
+    return hasse_invariant(local_symbol(G, p)) == hilbert_symbol(QQ(-1), -d, p)
+  elseif n == 4
+    return !is_local_square(d, p) || (hasse_invariant(local_symbol(G, p)) == hilbert_symbol(QQ(-1), QQ(-1), p))
+  else
+    return true
+  end
+end
+
+@doc raw"""
+    smallest_kneser_prime(G::ZZGenus) -> ZZRingElem
+
+Given a genus $G$ of integral $\mathbb{Z}$-lattice, return the
+smallest Kneser prime $p$ for $G$.
+
+A prime number $p$ is called Kneser for $G$ if:
+- The lattices in $G$ are isotropic locally at $p$;
+- The lattices in $G$ are maximal bilinear integral locally at $p$;
+- The prime $p$ is not an improper spinor generator of $G$.
```

Comment:
Please double check this. Maybe being a generator is too strong?

Reply:
I wanted to use the function, but it is not defined everywhere (for example at the bad primes). So maybe I should skip the bad primes, and then only check that `f(p) != 0` ?

Reply:
yes

Reply:
well in some cases the bad primes are interesting for us ... but then I am not sure how the spinor genus behaves. 
In case there is a unique one it is no problem anyways.